typeahead.js

```diff
@@ -0,0 +1,84 @@
+                'Winston-Salem', 'North Las Vegas', 'Gilbert', 'Chesapeake', 'Irving', 'Hialeah', 'Garland', 'Fremont', 'Richmond', 'Boise', 'Baton Rouge', 'Des Moines'];
+let suggestions;
+
+class Typeahead {
+    constructor(containerID) {
+        document.getElementById(containerID).innerHTML = '<input id=\'typeahead\' type=\'text\'></input><ul id=\'suggestions\'></ul>';
+        addEventListeners();
+    }
+}
+
+function addEventListeners() {
+    const searchBar = document.getElementById('typeahead');
+    const suggestionDisplay = document.getElementById('suggestions');
+
+    let HTMLSuggestions;
+    let firstOption;
+    let lastOption;
+
+    searchBar.addEventListener('input', (event) => {
+        const input = event.target.value;
+        if (input.length > 0) {
+            const regex = new RegExp(`^${input}`, 'i');
+            suggestions = statesList.sort().filter(word => regex.test(word));
+        } else {
+            suggestions = [];
+        }
+        displaySuggestions();
+        HTMLSuggestions = Array.from(document.querySelectorAll('li.suggestion'));
+        firstOption = HTMLSuggestions[0];
+        lastOption = HTMLSuggestions[HTMLSuggestions.length - 1];
+        toggleHighlight(firstOption)
+    });
+
+    suggestionDisplay.addEventListener('click', (event) => {
+        searchBar.value = event.target.textContent;
+        suggestions = [];
+        displaySuggestions();
+    });
+
+    suggestionDisplay.addEventListener('mouseover', (event) => {
+        if (event.target !== firstOption) {
+            toggleHighlight(event.target);
+            if (firstOption.id === 'highlight') {
+                toggleHighlight(firstOption);
+            }
+        } else if (firstOption.id !== 'highlight') {
+            toggleHighlight(firstOption);
+        }
+    });
+
+    suggestionDisplay.addEventListener('mouseout', (event) => {
+        toggleHighlight(event.target);
+    });
+
+    searchBar.addEventListener('keydown', (event) => {
+        const previousOption = document.getElementById('highlight');
+        if (event.keyCode === 40) {
+            event.preventDefault();
+            toggleHighlight(previousOption);
+            previousOption.nextSibling !== null ? toggleHighlight(previousOption.nextSibling) : toggleHighlight(firstOption);
+        } else if (event.keyCode === 38) {
+            event.preventDefault();
+            toggleHighlight(previousOption);
+            previousOption.previousSibling !== null ? toggleHighlight(previousOption.previousSibling) : toggleHighlight(lastOption);
+        } else if (event.keyCode === 13) {
+            event.preventDefault();
+            searchBar.value = previousOption.textContent;
+        }
+    });
+}
+
+function displaySuggestions() {
+    document.getElementById('suggestions').innerHTML = (suggestions.map(element => '<li class=\'suggestion\'>'+element+'</li>')).join('');
```

Comment:
How about using Array.prototype.reduce here? It should be able to combine the work of .map and .join